src/stan_math_backend/Expression_gen.ml

```diff
@@ -567,7 +567,10 @@ and pp_expr ppf Expr.Fixed.({pattern; meta} as e) =
           Expr.Typed.(local_scalar (type_of t) (adlevel_of t))
           pp_expr e
       in
-      let tform ppf = pf ppf "(@[<hov 2>@,%a@ ?@ %a@ :@ %a@])" in
+      let tform ppf =
+        pf ppf
+          "(@[<hov 2>@,%a@ ?@ stan::math::eval(%a)@ :@ stan::math::eval(%a)@])"
+      in
       if types_match et ef then tform ppf pp_expr ec pp_expr et pp_expr ef
       else tform ppf pp_expr ec promoted (e, et) promoted (e, ef)
```

Comment:
Beat me to it! How about something like this

```ocaml
      let tform ppf = pf ppf "(@[<hov 2>@,%a@ ?@ %a@ :@ %a@])" in
      let eval_pp ppf a =
        if UnsizedType.contains_eigen_type meta.type_ then
          pf ppf "stan::math::eval(%a)" pp_expr a
        else pf ppf "%a" pp_expr a
      in
      if types_match et ef then tform ppf pp_expr ec eval_pp et eval_pp ef
      else tform ppf eval_pp ec promoted (e, et) promoted (e, ef)
```

So that if the statement doesn't contain an eigen type we don't bother having to call `eval()`

Reply:
Did you mean `UnsizedType.is_eigen_type`? If so I've done this exactly